File: dp/c++/luggage.cpp
#include <bits/stdc++.h>

using namespace std;

int memo[25][105];
vector<int> p;

int fr(int i, int w) {

  if (w == 0)
    return true;
  if (w < 0 || i == (int)p.size())
    return false;

  if (memo[i][w] != -1)
    return memo[i][w];

  int res = fr(i + 1, w - p[i]) || fr(i + 1, w);
  return memo[i][w] = res;
}

bool check_parity() {
  int sum = 0;

  for (int peso : p) {
    sum += peso;
  }

  if (sum % 2 != 0)
    return false;

  memset(memo, -1, sizeof(memo));
  return fr(0, sum / 2);
}

int main() {

  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int tc;
  if (!(cin >> tc))
    return 0;

  string line;

  getline(cin, line);

  while (tc--) {

    getline(cin, line);

    stringstream ss(line);
    int peso;

    p.clear();

    while (ss >> peso) {
      p.push_back(peso);
    }

    if (check_parity()) {
      cout << "YES\n";
    } else {
      cout << "NO\n";
    }
  }

  return 0;
}

File: dp/python/luggage.py
from sys import stdin as sd


def check_parity(p: list[int]):

    w = sum(p)
    if w % 2 == 1:
        return False
    return fr(0, w // 2, {}, p)


def fr(i: int, w: int, memo: dict[int, int], p: list[int]):

    if w == 0:
        return True
    if w < 0:
        return False
    if i == len(p):
        return False

    if (i, w) in memo:
        return memo[(i, w)]

    res = fr(i + 1, w - p[i], memo, p) or fr(i + 1, w, memo, p)
    memo[(i, w)] = res
    return res


def main(sd):

    tc = int(sd.readline().strip())
    for _ in range(tc):
        line = sd.readline().split()

        p = list(map(int, line))

        if check_parity(p):
            print("YES")
        else:
            print("NO")


main(sd)
